Implement MIN/MAX AMBIENT safety checks

wavexx opened the pull request:
Take advantage of the NTC thermistor found on the Einsy as an additional
safety measure, following the steps of the other MIN/MAXTEMP errors.

Introduce two configurable params AMBIENT_MINTEMP and AMBIENT_MAXTEMP
in the variant defines and set them for the MK3/MK3S to -30/+100
respectively.

AMBIENT_MINTEMP is primarily intended to catch a defective board
thermistor (to ensure MAXTEMP would be properly triggered) and thus the
trigger temperature is set just above the sensing limit and well below
the operating range.

AMBIENT_MAXTEMP is set at 100C, which is instead 20C above the maximum
recommended operating temperature of the Einsy. The NTC thermistor is
located just above the main power connector on the bottom of the board,
and could also help in detecting a faulty connection which can result in
rapid overheating of the contacts.

As for MAXTEMP, we cut power to the heaters, print fan and motors to
reduce power draw. Resume is not possible except by resetting the
printer, since the user is highly advised to inspect the board for
problems before attempting to continue.

PFW-1114

wavexx committed:
Remove bogus comment (BED_MINTEMP *is* implemented)

wavexx committed:
Remove useless assignment

target_temperature_bed is already reset by disable_heaters() in Stop()

wavexx committed:
Remove redundant disable_heater() calls in max/min_temp handling

In max/min_temp handlers remove the redundant disable_heater() call.
Handlers already need to call Stop(), which will disable all heaters
as the first step.

Fix comments in order to mention that all heaters get disabled.
Use "MAX/MINTEMP BED" correctly in both the LCD and serial.

wavexx committed:
Add some important notes about thermistor ADC handling

wavexx committed:
Implement MIN/MAX AMBIENT safety checks

Take advantage of the NTC thermistor found on the Einsy as an additional
safety measure, following the steps of the other MIN/MAXTEMP errors.

Introduce two configurable params AMBIENT_MINTEMP and AMBIENT_MAXTEMP
in the variant defines and set them for the MK3/MK3S to -30/+100
respectively.

AMBIENT_MINTEMP is primarily intended to catch a defective board
thermistor (to ensure MAXTEMP would be properly triggered) and thus the
trigger temperature is set just above the sensing limit and well below
the operating range.

AMBIENT_MAXTEMP is set at 100C, which is instead 20C above the maximum
recommended operating temperature of the Einsy. The NTC thermistor is
located just above the main power connector on the bottom of the board,
and could also help in detecting a faulty connection which can result in
rapid overheating of the contacts.

As for MAXTEMP, we cut power to the heaters, print fan and motors to
reduce power draw. Resume is not possible except by resetting the
printer, since the user is highly advised to inspect the board for
problems before attempting to continue.

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -1484,32 +1504,24 @@ void min_temp_error(uint8_t e) {
   if(IsStopped() == false) {
     SERIAL_ERROR_START;
-    SERIAL_ERRORLNPGM("Temperature heated bed switched off. MAXTEMP triggered !");
+    SERIAL_ERRORLNPGM("Heaters switched off. MAXTEMP BED triggered !");
```
Can we build these strings from parameters? There are multiple instances of them and only MAXTEMP/ MINTEMP BED/AMBIENT actually change.

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -1428,19 +1452,16 @@ enum { LCDALERT_NONE = 0, LCDALERT_HEATERMINTEMP, LCDALERT_BEDMINTEMP, LCDALERT_
 uint8_t last_alert_sent_to_lcd = LCDALERT_NONE;
 
 void max_temp_error(uint8_t e) {
-  disable_heater();
```
I'd rather keep the disable_heater() call here - just to be safe

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -1459,12 +1480,11 @@ void min_temp_error(uint8_t e) {
 	return;
 #endif
 //if (current_temperature_ambient < MINTEMP_MINAMBIENT) return;
-  disable_heater();
```
also keep the disable_heater() here

wavexx replied:
Stop() does that of course as the first call. Maybe we should move Stop() as the first action instead?

DRracer reviewed (changes requested):
Please address my comments in the code

wavexx commented:
Instead of adding a duplicate call, what about moving Stop() first, which makes more sense anyway?
Stop() will disable heaters in the first step too.

wavexx committed:
Re-introduce redundant disable_heaters() calls

Partially revert 285b505c73a54e9af01816e3a614de73ad181851 so that
we ensure heaters are disabled ASAP in case of potential bugs
in the max_*_error functions.

wavexx committed:
Avoid redundant temperature error strings

Factor-out MIN/MAXTEMP [BED/AMB] out of the error message, which is now
built at runtime instead.

Introduce two missing ultralcd functions lcd_setalertstatus and
lcd_updatestatus to handle regular strings.

246272 -> 246084 = 188 bytes saved

wavexx commented:
Rebased on current master. I re-introduced the disabled_heaters calls, as well as building the error strings out of parameters. Despite two new small lcd functions, still saves 188 bytes.

There's probably a bit more to save during the initialization of the minmax/temp[raw] tables IMHO.

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -1427,20 +1451,47 @@ enum { LCDALERT_NONE = 0, LCDALERT_HEATERMINTEMP, LCDALERT_BEDMINTEMP, LCDALERT_
+void temp_update_messagepgm(const char* PROGMEM type, void (*func)(const char*) = lcd_updatestatus)
+{
+    char msg[LCD_WIDTH];
+    strcpy_P(msg, PSTR("Err: "));
```
interestingly, filling the msg this way results in 30B shorter code than in `char msg[] = "Err: ";`

wavexx replied:
Tried a few different approaches, I was surprised too. It's a bit of a gimmick to pass the update function, but I didn't want to duplicate the str* function twice.

When I was looking to unify some code I think we could actually also remove the "Err: " prefix in a few other places by adding a proper lcd function to set error messages. But it had a bit too many changes for this PR. Something to think about for the future though.

DRracer commented on Firmware/temperature.cpp:
```diff
@@ -1484,44 +1532,61 @@ void min_temp_error(uint8_t e) {
-#endif
-	static const char err[] PROGMEM = "Err: MINTEMP BED";
+    disable_heater();
+	static const char err[] PROGMEM = "MINTEMP BED";
```
indentation inconsistent with lines above

wavexx replied:
I think it's just the usual tab/spaces expanding differently yet again, they're aligned in the code. Expanding it now.

DRracer reviewed (changes requested):
Please fix the indentation inconsistency at one spot, then we can merge this.

wavexx committed:
Fix indentation
